src/api/artifact/controller.go

```diff
@@ -54,10 +60,220 @@ type Controller interface {
 }
 
 // NewController creates an instance of the default artifact controller
-func NewController() Controller {
-	// TODO implement
+func NewController(artMgr artifact.Manager, tagMgr tag.Manager) Controller {
+	return &controller{
+		artMgr: artMgr,
+		tagMgr: tagMgr,
+	}
+}
+
+// TODO handle concurrency, the redis lock doesn't cover all cases
+// TODO As a redis lock is applied during the artifact pushing, we do not to handle the concurrent issues
+// for artifacts and tags？？
+
+type controller struct {
+	artMgr artifact.Manager
+	tagMgr tag.Manager
+}
+
+func (c *controller) Ensure(ctx context.Context, repository *models.RepoRecord, digest string, tags ...string) error {
+	artifact, err := c.ensureArtifact(ctx, repository, digest)
+	if err != nil {
+		return err
+	}
+	for _, tag := range tags {
+		if err = c.ensureTag(ctx, artifact, tag); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ensure the artifact exists under the repository, create it if doesn't exist.
+func (c *controller) ensureArtifact(ctx context.Context, repository *models.RepoRecord, digest string) (*artifact.Artifact, error) {
+	query := &q.Query{
+		Keywords: map[string]interface{}{
+			"repository_id": repository.RepositoryID,
+			"digest":        digest,
+		},
+	}
+	_, artifacts, err := c.artMgr.List(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	// the artifact already exists under the repository, return directly
+	if len(artifacts) > 0 {
+		return artifacts[0], nil
+	}
+
+	// the artifact doesn't exist under the repository, create it first
+	artifact := &artifact.Artifact{
+		ProjectID:    repository.ProjectID,
+		RepositoryID: repository.RepositoryID,
+		Digest:       digest,
+		PushTime:     time.Now(),
+	}
+	// abstract the specific information for the artifact
+	c.abstract(ctx, artifact)
+	// create it
+	id, err := c.artMgr.Create(ctx, artifact)
```

Comment:
Concurrent case.  It should handle the dup error?

Reply:
Firstly, ideally, there shouldn't be concurrent issue as we introduce lock in quota middleware.
Secondly, even though got dup error, the error is wrapped in dao level and shouldn't be handled here

Reply:
1. I don't think as a public func you should assume the caller's behavior.
2. Not returning dup error is business logic that should not be handled by dao level.  And after quick checking I don't see dao level eating dup error.

Reply:
Searching `IsConflictError ` for the wrap of dup error. 
I think this is one of the most useful function of the common error we designed? Wrap the error with specific error code and detect the code in the API level

Reply:
The `ensureArtifact` means must ensure that artifact referenced by (repository_id, digest) exists in db. No matter the artifact exists before we call this method, or in the method we found the artifact not exist and before we create it but it was created by others.

We needed to deal with dup error in this method. I think this method is very similar with `get_or_create` of `Django` method. The following is the implemention of `get_or_create` in `Django`.

```python
    def get_or_create(self, defaults=None, **kwargs):
        """
        Look up an object with the given kwargs, creating one if necessary.
        Return a tuple of (object, created), where created is a boolean
        specifying whether an object was created.
        """
        # The get() needs to be targeted at the write database in order
        # to avoid potential transaction consistency problems.
        self._for_write = True
        try:
            return self.get(**kwargs), False
        except self.model.DoesNotExist:
            params = self._extract_model_params(defaults, **kwargs)
            return self._create_object_from_params(kwargs, params)

    def _create_object_from_params(self, lookup, params, lock=False):
        """
        Try to create an object using passed params. Used by get_or_create()
        and update_or_create().
        """
        try:
            with transaction.atomic(using=self.db):
                params = dict(resolve_callables(params))
                obj = self.create(**params)
            return obj, True
        except IntegrityError as e:
            try:
                qs = self.select_for_update() if lock else self
                return qs.get(**lookup), False
            except self.model.DoesNotExist:
                pass
            raise e
```

First, it try to get the instance, then create it when not found and get again if create failed.

Reply:
Done